OSDOCS-19870: adds main console docs RHCL (#112113)
openshift/openshift-docs: modules/ref-rhcl-api-management-rbac.adoc

API management enforces strict namespace boundaries:

Consumer namespaces:: Consumers manage `APIKey` and secret resources in their assigned namespaces only.
Owner namespaces:: Owners manage `APIProduct` and `APIKeyApproval` resources in their team namespaces.
Cross-namespace references:: `APIKey` resources can reference `APIProduct` resources in other namespaces, but consumers can create `APIKey` resources only in namespaces where they have role bindings.

API key values are protected through architectural isolation by the following workflow:

. The API consumer creates a Secret containing the API key value, then creates an `APIKey` resource referencing that Secret.
. The controller creates an `APIKeyRequest` shadow resource in the owner namespace and does not contain the API key value.
. The API owner creates an `APIKeyApproval` resource in their namespace.
. The controller creates an enforcement Secret in the {prodname} namespace, which the API consumer cannot access.
. API owners do not see API consumer key values. Even administrators do not have Secret-read permissions in consumer namespaces.

[id="ref-rhcl-api-management-cluster-roles_{context}"]
== Cluster roles

> {prodname} provides four predefined cluster roles for API management. When configuring `ClusterRole` objects, make sure that include all of the information you need in a single role. Do not name two roles identically to add or change permissions. `ClusterRole` objects with the same `metadata.name` parameter value conflict.

Reviewer: ```suggestion
{prodname} provides four predefined cluster roles for API management. When you configure `ClusterRole` objects, include all of the information you need in a single role. Do not name two roles identically to add or change permissions. `ClusterRole` objects with the same `metadata.name` parameter value conflict.
```

The `api-catalog-browser` role provides cluster-wide read access to API catalog resources for discovery. This role is used by API browsers and permissions are also included in the API owner and API administrator roles.
+

> .Example api-catalog-browser permissions

Reviewer: Double check the preview

<img width="792" height="284" alt="Image" src="https://github.com/user-attachments/assets/8c187d30-d87b-4e8b-a49e-b1b6c1d64b64" />

[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: api-catalog-browser
rules:
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apiproducts"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["extensions.kuadrant.io"]
    resources: ["planpolicies"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["kuadrant.io"]
    resources: ["authpolicies", "ratelimitpolicies"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["gateway.networking.k8s.io"]
    resources: ["httproutes", "gateways"]
    verbs: ["get", "list", "watch"]
----

[IMPORTANT]
====
The `api-catalog-browser` cluster role binding grants cluster-wide read access. There is no way to limit catalog browsing to specific namespaces. All catalog browsers see all published products.
====

The `api-consumer` role enables API consumers to create and manage API keys in assigned namespaces. Bind this role in API consumer namespaces by using a `RoleBinding` resource.

> +

Reviewer: Check the preview here, too. Looks a little wonky.

.Example api-consumer permissions
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: api-consumer
rules:
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeys"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: [""]
    resources: ["secrets"]
    verbs: ["get", "list", "watch", "create", "update", "delete"]
----

> The `api-owner role` lets teams to publish APIs and manage consumer access requests. This role has the following characteristics:

Reviewer: ```suggestion
With the `api-owner` role, you can publish APIs and manage consumer access requests. This role has the following characteristics:
```

* Owners can view `APIKeyRequest` custom resources (CRs), but not API consumer `APIKey` or Secret resources.

.Example api-owner role permissions
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: api-owner
rules:
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apiproducts"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeyapprovals"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeyrequests"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["extensions.kuadrant.io"]
    resources: ["planpolicies"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["kuadrant.io"]
    resources: ["authpolicies", "ratelimitpolicies"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["gateway.networking.k8s.io"]
    resources: ["httproutes", "gateways"]
    verbs: ["get", "list", "watch"]
----

The `api-admin` role gives cluster administrators cluster-wide access for management and troubleshooting. This role has the following characteristics:

> * Same core permissions as `api-owner` role but cluster-wide

Reviewer: ```suggestion
* Same core permissions as `api-owner` role, but cluster-wide
```

* Additional `APIKey` write access for troubleshooting
* Intentional limitation: No secret read permissions in consumer namespaces

.Example api-admin namespace-scoped role permissions
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: api-admin
rules:
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apiproducts"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeyapprovals"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeyrequests"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeys"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: ["devportal.kuadrant.io"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["kuadrant.io"]
    resources: ["authpolicies", "ratelimitpolicies"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["gateway.networking.k8s.io"]
    resources: ["httproutes", "gateways"]
    verbs: ["get", "list", "watch"]
----

> * With `api-owner` permissions, cluster-wide are granted with the `apiGroups` `"devportal.kuadrant.io"` value and the `"apiproducts"` `"apikeyapprovals"` `"apikeyrequests"` `resources` and all available actions.

Reviewer: ```suggestion
* With `api-owner` permissions, you gain cluster-wide permissions and can complete all available actions with the `apiGroups` `"devportal.kuadrant.io"` value and the `"apiproducts"` `"apikeyapprovals"` `"apikeyrequests"` `resources`.
```
Feel free to revise if my suggestion is wrong.

> * Additional troubleshooting with full `APIKey` access is granted with the `apiGroups` `"devportal.kuadrant.io"` value and the `"apikeys"` `resources` and all available actions..

Reviewer: ```suggestion
* With the `apiGroups` `"devportal.kuadrant.io"` value and the `"apikeys"` `resources`, you can complete additional troubleshooting tasks with full `APIKey` access and take all available actions.
```
Not sure if my revision is correct.

> * Catalog resources are granted with all apiGroups assigned, such as `"devportal.kuadrant.io"`, `"kuadrant.io"`, and `"gateway.networking.k8s.io"` but with the limited actions, `"get", "list", "watch" `.

Reviewer: ```suggestion
* Catalog resources are granted with all apiGroups assigned, such as `"devportal.kuadrant.io"`, `"kuadrant.io"`, and `"gateway.networking.k8s.io"`, but you are limited to the `"get"`, `"list"`, and  `"watch"` actions.
```
Again, not sure if my revision is accurate.

[id="ref-rhcl-api-management-role-binding-requirements_{context}"]
== Role binding requirements

> 

Reviewer: It would be good to have an introductory sentence here.

    resources: ["apikeys"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: [""]
    resources: ["secrets"]
    verbs: ["get", "list", "watch", "create", "update", "delete"]
----

The `api-owner role` lets teams to publish APIs and manage consumer access requests. This role has the following characteristics:

* Includes cluster-wide read access for catalog browsing.
* Namespace-scoped write access for `APIProduct` and `APIKeyApproval` custom definition resources (CRDs).
* Owners can view `APIKeyRequest` custom resources (CRs), but not API consumer `APIKey` or Secret resources.

.Example api-owner role permissions
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: api-owner
rules:
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apiproducts"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeyapprovals"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeyrequests"]
    verbs: ["get", "list", "watch"]
---

> apiVersion: rbac.authorization.k8s.io/v1

Reviewer 2: I still see two clusterroles. it should only be one.

. API owners do not see API consumer key values. Even administrators do not have Secret-read permissions in consumer namespaces.

[id="ref-rhcl-api-management-cluster-roles_{context}"]
== Cluster roles

{prodname} provides the following four predefined cluster roles for API management:

The `api-catalog-browser` role provides cluster-wide read access to API catalog resources for discovery. This role is used by API browsers and permissions are also included in the API owner and API administrator roles.
+
.Example api-catalog-browser permissions
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: api-catalog-browser
rules:
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apiproducts"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["extensions.kuadrant.io"]
    resources: ["planpolicies"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["kuadrant.io"]
    resources: ["authpolicies", "ratelimitpolicies"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["gateway.networking.k8s.io"]
    resources: ["httproutes", "gateways"]
    verbs: ["get", "list", "watch"]
----

> [IMPORTANT]

Reviewer 2: The cluster role by itself does not grants cluster wide access. It's the binding actually granting access.

  - apiGroups: [""]
    resources: ["secrets"]
    verbs: ["get", "list", "watch", "create", "update", "delete"]
----

The `api-owner role` lets teams to publish APIs and manage consumer access requests. This role has the following characteristics:

* Includes cluster-wide read access for catalog browsing.
* Namespace-scoped write access for `APIProduct` and `APIKeyApproval` custom definition resources (CRDs).
* Owners can view `APIKeyRequest` custom resources (CRs), but not API consumer `APIKey` or Secret resources.

.Example api-owner namespace-scoped role permissions
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: api-owner
rules:
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apiproducts"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeyapprovals"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeyrequests"]
    verbs: ["get", "list", "watch"]
----

.Example api-owner cluster-wide-scoped role permissions for discovery
[source,yaml]
----

> apiVersion: rbac.authorization.k8s.io/v1

Reviewer 2: Do not break the clusterrole resource into two parts. The api-owner permissions should include both. Additionally, you cannot have two cluster roles with the same name. They conflict.

    verbs: ["get", "list", "watch"]
  - apiGroups: ["devportal.kuadrant.io"]
    resources: ["apikeys"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
  - apiGroups: ["devportal.kuadrant.io"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["kuadrant.io"]
    resources: ["authpolicies", "ratelimitpolicies"]
    verbs: ["get", "list", "watch"]
  - apiGroups: ["gateway.networking.k8s.io"]
    resources: ["httproutes", "gateways"]
    verbs: ["get", "list", "watch"]
----

* With `api-owner` permissions, cluster-wide are granted with the `apiGroups` `"devportal.kuadrant.io"` value and the `"apiproducts"` `"apikeyapprovals"` `"apikeyrequests"` `resources` and all available actions.
* Additional troubleshooting with full `APIKey` access is granted with the `apiGroups` `"devportal.kuadrant.io"` value and the `"apikeys"` `resources` and all available actions..
* Catalog resources are granted with all apiGroups assigned, such as `"devportal.kuadrant.io"`, `"kuadrant.io"`, and `"gateway.networking.k8s.io"` but with the limited actions, `"get", "list", "watch" `.

[id="ref-rhcl-api-management-role-binding-requirements_{context}"]
== Role binding requirements

> API browser binding:: API browsers require one cluster-role binding for `api-catalog-browser` to enable cluster-wide catalog discovery.

Reviewer 2: We do not define an "API Browser" persona. The API Consumer persona requires two role bindings: the "api-browser" role at cluster-wide scope and the "api-consumer" role at namespace scope.

Author: Hmm, we do, according to the intro: browser, consumer, owner, admin. Did these change?

Reviewer 2: We define "API Browsing cluster role" that needs to be assigned (bound) to any user (or group) playing the role of API Consumer in a cluster wide mode. Thus,

API Consumer gets two cluster role bounds:
* api-consumer (namespace scoped)
* api-browser (cluster wide scoped)
